packages/rocketchat-statistics/server/lib/SAUMonitor.js

```diff
@@ -0,0 +1,403 @@
+			callback(session.connectionHandle);
+		});
+	}
+
+	_applyAllServerSessionsIds(callback) {
+		if (!callback || typeof callback !== 'function') {
+			return;
+		}
+
+		const sessions = Object.values(Meteor.server.sessions);
+		const sessionIds = sessions.map((s) => s.id);
+		while (sessionIds.length) {
+			callback(sessionIds.splice(0, 500));
+		}
+	}
+
+	_updateConnectionInfo(sessionId, data = {}) {
+		if (!sessionId) {
+			return;
+		}
+		if (Meteor.server.sessions[sessionId]) {
+			Object.keys(data).forEach((p) => {
+				Object.defineProperty(Meteor.server.sessions[sessionId].connectionHandle, p, {
+					value: data[p],
+				});
+			});
+		}
+	}
+
+	_applyAllServerSessionsBatch(callback, params) {
+		const batch = (arr, limit) => {
+			if (!arr.length) {
+				return Promise.resolve();
+			}
+			const ids = [];
+			return Promise.all(arr.splice(0, limit).map((item) => {
+				ids.push(item.id);
+				return this._getConnectionInfo(item.connectionHandle, params);
+			})).then((data) => {
+				callback(data, ids);
+				return batch(arr, limit);
+			}).catch((e) => {
+				logger.debug(`Error: ${ e.message }`);
+				this._log(`${ this._serviceName } - Error: ${ e.message }`);
+			});
+		};
+
+		const sessions = Object.values(Meteor.server.sessions);
+		batch(sessions, 500);
+	}
+
+	_startAggregation() {
+		logger.info('[aggregate] - Start Cron.');
+		SyncedCron.add({
+			name: 'aggregate-sessions',
+			schedule: (parser) => parser.text('every 10 seconds'),
+			job: () => {
+				this.aggregate();
+			},
+		});
+
+		SyncedCron.start();
+	}
+
+	aggregate() {
+		logger.info('[aggregate] - Aggregatting data.');
+
+		const date = new Date();
+		date.setDate(date.getDate() - 1); // yesterday
+		const yesterday = getDateObj(date);
+
+		const match = {
+			userId: { $exists: true },
+			lastActivityAt: { $exists: true },
+			device: { $exists: true },
+			type: 'session',
+			year: { $lte: yesterday.year },
+			month: { $lte: yesterday.month },
+			day: { $lte: yesterday.day },
+		};
+
+		Sessions.model.rawCollection().aggregate([{
+			$match: match,
+		}, {
+			$group: {
+				_id: {
+					userId: '$userId',
+					day: '$day',
+					month: '$month',
+					year: '$year',
+				},
+				times: { $push: { $trunc: { $divide: [{ $subtract: ['$lastActivityAt', '$loginAt'] }, 1000] } } },
+				devices: { $addToSet: '$device' },
+			},
+		}, {
+			$project: {
+				_id: '$_id',
+				times: { $filter: { input: '$times', as: 'item', cond: { $gt: ['$$item', 0] } } },
+				devices: '$devices',
+			},
+		}, {
+			$project: {
+				type: 'user_daily',
+				day: '$_id.day',
+				month: '$_id.month',
+				year: '$_id.year',
+				userId: '$_id.userId',
+				time: { $sum: '$times' },
+				count: { $size: '$times' },
+				devices: '$devices',
+			},
+		}]).forEach(Meteor.bindEnvironment((record) => {
+			record._id = `${ record.userId }-${ record.year }-${ record.month }-${ record.day }`;
+			Sessions.upsert({ _id: record._id }, record);
+		}));
+
+		Sessions.update(match, {
```

Comment:
why do you update records here instead of upadating/inserting it on the [upsert](https://github.com/RocketChat/Rocket.Chat/pull/11525/files#diff-391ef8ba5639084906847ad4e9579ab8R393) above?

Reply:
Will be difficult to update different records since we are filtering by records of yesterday